fix/ota-3634/Change selfupdate to use fake package manager

Zee314159 opened the pull request:
Draft code to make sure I'm on the right track. Still in progress. There are some directory related errors. Checking the installed file appeared in the installation directory has not finished yet.

patrickvacek commented on src/libaktualizr/package_manager/CMakeLists.txt:
```diff
@@ -13,6 +13,7 @@ aktualizr_source_file_checks(${SOURCES} packagemanagerconfig.cc ${HEADERS})
 target_sources(config PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/packagemanagerconfig.cc)
 
 add_aktualizr_test(NAME packagemanagerfake SOURCES packagemanagerfake_test.cc LIBRARIES PUBLIC uptane_generator_lib)
+add_test(NAME test_install_aktualizr_and_update COMMAND ${PROJECT_SOURCE_DIR}/scripts/test_install_aktualizr_and_update.sh ${PROJECT_BINARY_DIR})
```
I'd suggest moving this line to `tests/CMakeLists.txt`. That's the generic place for miscellaneous tests, and this seems like an unexpected location to me.

patrickvacek commented on tests/config/selfupdate.toml:
```diff
@@ -16,4 +16,5 @@ primary_ecu_serial = "723f79763eda1c753ce565c16862c79acdde32eb922d6662f088083c51
 
 [pacman]
-type = "debian"
+type = "none"
```
Not a big deal or even necessary, but we might consider renaming these `selfupdate` bits if they no longer with be actually self-updating.

Zee314159 replied:
So what might be a better name to replace ‘selfupdate'? fakeupdate or testupdate?

patrickvacek replied:
`testupdate` is okay. I'm trying to think of something better, but I mostly just want to get `self` out of there since it isn't accurate, and `update` on it's own is obviously too broad. :)

patrickvacek commented on scripts/test_install_aktualizr_and_update.sh:
```diff
@@ -2,28 +2,26 @@
+# check the updated file appeared in the installation directory
+#folder="$TEST_INSTALL_DESTDIR/"
+#if [ ! -d "$folder"]; then
+#fi
```
The downloaded and "installed" file should show up something in the storage directory, and we should verify that it is there and maybe that the sha256sum matches our expectations.

Zee314159 commented:
`130/130 Test #129: test_install_aktualizr_and_update ...................***Timeout 1500.00 sec
++ mktemp -d
+ TEST_INSTALL_DESTDIR=/tmp/tmp.BX8KkYCf0S/aktualizr
+ /builds/olp/edge/ota/connect/client/aktualizr/build-coverage/src/aktualizr_primary/aktualizr --version
+ ./scripts/selfupdate_server.py 8000 /tmp/tmp.BX8KkYCf0S/aktualizr
Current aktualizr version is: 2019.8-9-g86b24446
+ /builds/olp/edge/ota/connect/client/aktualizr/build-coverage/src/uptane_generator/uptane-generator generate --path /tmp/tmp.BX8KkYCf0S/aktualizr/fake_root
Uptane metadata repos generated at "/tmp/tmp.BX8KkYCf0S/aktualizr/fake_root"
+ /builds/olp/edge/ota/connect/client/aktualizr/build-coverage/src/uptane_generator/uptane-generator image --path /tmp/tmp.BX8KkYCf0S/aktualizr/fake_root --targetname selfupdate_2.0 --filename ./tests/test_data/selfupdate_2.0 --hwid selfupdate-device
Added a target "selfupdate_2.0" to the images metadata
+ /builds/olp/edge/ota/connect/client/aktualizr/build-coverage/src/uptane_generator/uptane-generator addtarget --path /tmp/tmp.BX8KkYCf0S/aktualizr/fake_root --targetname selfupdate_2.0 --hwid selfupdate-device --serial 723f79763eda1c753ce565c16862c79acdde32eb922d6662f088083c51ffde66
Added target selfupdate_2.0 to director targets metadata for ECU with serial 723f79763eda1c753ce565c16862c79acdde32eb922d6662f088083c51ffde66 and hardware ID selfupdate-device
+ /builds/olp/edge/ota/connect/client/aktualizr/build-coverage/src/uptane_generator/uptane-generator signtargets --path /tmp/tmp.BX8KkYCf0S/aktualizr/fake_root
Signed the staged director targets metadata
++ mktemp -d
+ TEMP_DIR=/tmp/tmp.fBXgsGTk0r
+ mkdir -m 700 -p /tmp/tmp.fBXgsGTk0r/import
+ cp ./tests/test_data/prov_selfupdate/ca.pem ./tests/test_data/prov_selfupdate/client.pem ./tests/test_data/prov_selfupdate/device_id ./tests/test_data/prov_selfupdate/ecukey.pem ./tests/test_data/prov_selfupdate/ecukey.pub ./tests/test_data/prov_selfupdate/is_registered ./tests/test_data/prov_selfupdate/misconfigured_ecus ./tests/test_data/prov_selfupdate/pkey.pem ./tests/test_data/prov_selfupdate/secondaries_list /tmp/tmp.fBXgsGTk0r/import
+ echo -e '[storage]\npath = "/tmp/tmp.fBXgsGTk0r"\n[import]\nbase_path = "/tmp/tmp.fBXgsGTk0r/import"'
+ /builds/olp/edge/ota/connect/client/aktualizr/build-coverage/src/aktualizr_primary/aktualizr -c ./tests/config/selfupdate.toml -c /tmp/tmp.fBXgsGTk0r/conf.toml once
Aktualizr version 2019.8-9-g86b24446 starting
Reading config: "/tmp/tmp.fBXgsGTk0r/conf.toml"
Reading config: "./tests/config/selfupdate.toml"
Bootstrap empty SQL storage
Bootstraping DB to version 20
Could not find primary ecu serial, defaulting to empty serial: unknown error
Could not find hardware_id for serial : unknown error
ECUs have been successfully registered to the server.
got SendDeviceDataComplete event
Current version for ECU ID: 723f79763eda1c753ce565c16862c79acdde32eb922d6662f088083c51ffde66 is unknown
got new updates
1 new update found in Uptane metadata.
Current version for ECU ID: 723f79763eda1c753ce565c16862c79acdde32eb922d6662f088083c51ffde66 is unknown
No file 'selfupdate_2.0' with matched hash in the database
got DownloadTargetComplete event
got AllDownloadsComplete event
Current version for ECU ID: 723f79763eda1c753ce565c16862c79acdde32eb922d6662f088083c51ffde66 is unknown
got InstallStarted event
Installing package using fake package manager
got InstallTargetComplete event
got AllInstallsComplete event
got PutManifestComplete event`

 `**Could not find primary ecu serial, defaulting to empty serial: unknown error
Could not find hardware_id for serial : unknown error**`
Are these errors negligible, and what caused the timeout?

mike-sul commented:
> **Could not find primary ecu serial, defaulting to empty serial: unknown error Could not find hardware_id for serial : unknown error**
> Are these errors negligible, and what caused the timeout?

Yes,, they are negligible.
Hard to say what causes the timeout, perhaps some details in the bash script.

patrickvacek commented on ci/gitlab/.gitlab-ci.yml:
```diff
@@ -183,7 +183,7 @@ xenial-pkg-test:
   script:
     - ./scripts/test_install_garage_deploy.sh
-    - ./scripts/test_install_aktualizr_and_update.sh
+    - ./scripts/test_install_aktualizr.sh
```
Should we also run this script (`./scripts/test_install_aktualizr.sh`) as part of `bionic-pkg-test`?

patrickvacek commented on tests/test_install_aktualizr_and_update.sh:
```diff
@@ -0,0 +1,42 @@
+    exit 1
+else
+    cd $TEST_INSTALL_IMAGE_DIR
+    for filename in `ls`; do
```
We should know filename in advance based on the hash. I'm debating the best way to get the hash, though. It might be to run `aktualizr-info --director-target | jq <something>` or maybe there's a better way.

It'd also be better to avoid a `cd` in the script.

Zee314159 commented:
> What's the status with the test failure? Is that still a concern?

The python server is never killed which causes the timeout on CI. Laurent has helped me solving the issue.
